File: web/src/utils/model-ref.ts
import { createLogger } from '@/lib/logger'

const log = createLogger('model-ref')

export interface ModelRefValue {
  provider: string
  modelId: string
}

/**
 * Decode a stored ``MODEL_REF`` setting value into a provider/model pair.
 *
 * Accepts the canonical ``{provider, model_id}`` JSON the pickers write and a
 * bare model string (read as model-only, provider empty, so a picker prompts
 * for an explicit provider selection).
 */
export function decodeModelRef(value: string): ModelRefValue {
  const text = value.trim()
  if (!text) return { provider: '', modelId: '' }
  if (text.startsWith('{')) {
    try {
      const parsed: unknown = JSON.parse(text)
      if (typeof parsed === 'object' && parsed !== null) {
        const { provider, model_id: modelId } = parsed as Record<string, unknown>
        return {
          provider: typeof provider === 'string' ? provider : '',
          modelId: typeof modelId === 'string' ? modelId : '',
        }
      }
    } catch (err) {
      log.warn('decodeModelRef: value is not valid JSON, treating as model-only', err)
    }
  }
  return { provider: '', modelId: value }
}

/** Serialize a provider/model pair to the canonical stored ``MODEL_REF`` JSON. */
export function encodeModelRef(provider: string, modelId: string): string {
  return JSON.stringify({ provider, model_id: modelId })
}

/**
 * Re-encode any accepted ``MODEL_REF`` spelling into one canonical string.
 *
 * The backend serializes with ``json.dumps`` (which pads after ``:`` and
 * ``,``) while the dashboard uses ``JSON.stringify`` (which does not), so the
 * same logical reference reaches a picker as two different strings depending
 * on which side last wrote it. Both parse identically server-side, but a
 * picker that preselects by string identity would show no selection for the
 * other spelling, so every ref is normalised before it is compared.
 */
export function normalizeModelRef(value: string): string {
  const { provider, modelId } = decodeModelRef(value)
  if (!provider || !modelId) return value
  return encodeModelRef(provider, modelId)
}

File: src/synthorg/settings/model_ref.py
"""Shared model-reference value for model-assignment settings.

A ``SettingType.MODEL_REF`` setting stores a provider + model pair as
canonical JSON ``{"provider": ..., "model_id": ...}`` so the value is
unambiguous and validatable against the live provider catalogue, rather than
a bare model string whose provider is implicit (the source of a class of
"model not found on the resolved provider" runtime failures).

An empty value means "unset". A bare (non-JSON) model string is read as
model-only, with the provider left unset, so the dashboard picker can surface
the model and prompt for an explicit provider selection.
"""

import json
from typing import Final

from pydantic import BaseModel, ConfigDict, Field


class ModelRef(BaseModel):
    """A provider + model selection for a model-assignment setting."""

    model_config = ConfigDict(frozen=True, allow_inf_nan=False, extra="forbid")

    provider: str = Field(
        default="",
        description="Provider connection name; empty when unset.",
    )
    model_id: str = Field(
        default="",
        description="Model id within the provider; empty when unset.",
    )

    @property
    def is_bound(self) -> bool:
        """True iff both a provider and a model are selected."""
        return bool(self.provider.strip()) and bool(self.model_id.strip())


_EMPTY: Final[ModelRef] = ModelRef()


def parse_model_ref(value: str) -> ModelRef:
    """Parse a stored MODEL_REF setting value into a :class:`ModelRef`.

    Accepts three forms: empty (unset), canonical
    ``{"provider", "model_id"}`` JSON, and a bare model string (read as
    model-only, provider empty).

    Returns:
        The parsed :class:`ModelRef`; an unset ref for an empty value.
    """
    text = value.strip()
    if not text:
        return _EMPTY
    if text.startswith("{"):
        try:
            data = json.loads(text)
        except ValueError:
            # Unparseable JSON: fall back to the raw value as a model-only
            # value rather than discarding the assignment entirely.
            return ModelRef(model_id=value)
        if isinstance(data, dict):
            provider = data.get("provider")
            model_id = data.get("model_id")
            return ModelRef(
                # Trimmed, unlike ``model_id``: a provider name is a registry
                # key and nothing else, so ``"  openai  "`` and ``"openai"``
                # name one connection. Left raw, every consumer would have to
                # remember to strip before its own lookup, and the one that
                # forgot would report the operator's provider unregistered.
                provider=provider.strip() if isinstance(provider, str) else "",
                model_id=model_id if isinstance(model_id, str) else "",
            )
        return ModelRef(model_id=value)
    # A bare model string (e.g. "glm-5.2") carries no provider: read it as
    # model-only so the picker can prompt for a provider selection. The raw
    # value is preserved (not the classification-stripped ``text``) so a
    # downstream structural guard still sees any untrimmed / control content.
    return ModelRef(model_id=value)


def serialize_model_ref(ref: ModelRef) -> str:
    """Serialize a :class:`ModelRef` to the canonical stored JSON string.

    Returns:
        ``{"provider": ..., "model_id": ...}`` JSON.
    """
    return json.dumps({"provider": ref.provider, "model_id": ref.model_id})
